File: python/S2_SORTING/L1_BASIC/insertionsort.py
class sort:
    def __init__(self,arr):
        self.arr=arr

    def insertionsort(self):
        n=len(self.arr)
        for i in range(0,n):
            key=self.arr[i]
            j=i-1
            while(j>=0 and key<self.arr[j]):
                self.arr[j+1]=self.arr[j]
                j-=1
            self.arr[j+1]=key
                
            
        
if __name__=="__main__":
    n=int(input("enter size of array :"))
    arr=list(map(int,input("enter array element: ").split()))
    s=sort(arr)
    s.insertionsort()
    for i in arr:
        print(i,end=" ")

File: C/S2_SORTING/L1_BASIC/insertionsort.c
/*
Insertion Sort

Algorithm:
Pick each element and insert it into its correct position in the already sorted part.
Assume the first element is already sorted.

Time Complexity:
Worst Case: O(n^2)
Average Case: O(n^2)
Best Case: O(n)   // when the array is already sorted

Space Complexity:
O(1)   // in-place, no extra space used

Properties:
- In-place sorting
- Stable sorting
*/
#include<stdio.h>
void insertionsort(int arr[],int n)
{
    for(int i=1;i<n;i++)
    {
        int key=arr[i];
        int j=i-1;
        while(j>=0 &&key<arr[j])//if array is already sorted ,this loop will never execute and best case happens
        {
            arr[j+1]=arr[j];
            j--;
        }
        arr[j+1]=key;
    }
}
int main()
{
    int n;
    printf("enter size of array: ");
    scanf("%d",&n);
    int arr[n];
    printf("enter element of array: ");
    for(int i=0;i<n;i++)
    {
        scanf("%d",&arr[i]);
    }
    insertionsort(arr,n);
    for(int i=0;i<n;i++)
    {
        printf("%d ",arr[i]);
    }
    return 0;
}